apps/site/pages/en/learn/test-runner/mocking.md

```diff
@@ -0,0 +1,270 @@
+
+export function main() {
+  const f = foo();
+}
+```
+
+Here, `foo` is an "own code" dependency of `main`.
+
+#### Why
+
+For a true unit test of `main`, `foo` should be mocked: you're testing that `main` works, not that `main` + `foo` work (that's a different test).
+
+#### Why not
+
+Mocking `foo` can be more trouble than worth, especially when `foo` is simple, well-tested, and rarely updated.
+
+Others argue that not mocking `foo` is better because it's more authentic and increases coverage of `foo` (because `main`'s tests will also verify `foo`). This can, however, create noise: when `foo` breaks, a bunch of other tests will also break, so tracking down the problem is more tedious: if only the 1 test for the item ultimately responsible for the issue is failing, that's very easy to spot; whereas 100 tests failing creates a needle-in-a-haystack to find the real problem.
+
+### External code
+
+This is what you/your project does not control.
+
+```mjs displayName="your-project/main.mjs"
+import bar from 'bar';
+
+export function main() {
+  const f = bar();
+}
+```
+
+Here, `bar` is a node_module installed via npm (or yarn, etc).
+
+Uncontroversially, for unit tests, this should always be mocked. For component and integration tests, whether to mock depends on what it is.
+
+#### Why
+
+Verifying that someone else's code works with your code is not the role of a unit test (and their code should already be tested).
+
+#### Why not
+
+Sometimes, it's just not realistic to mock. For example, you would almost never mock react (the medicine would be worse than the ailment).
+
+### External system
+
+These are things like databases, environments (Chromium or Firefox for a web app, an operating system for a node app, etc), file systems, memory store, etc.
+
+Ideally, mocking these would not be necessary. Aside from somehow creating isolated copies for each case (usually very impractical, due to cost, additional execution time, etc), the next best option is to mock. Without doing, tests sabotage each other:
+
+```mjs displayName="storage.mjs"
+import { db } from 'db';
+
+export function read(key, all = false) {
+  validate(key, val);
+
+  if (all) return db.getAll(key);
+
+  return db.getOne(key);
+}
+
+export function save(key, val) {
+  validate(key, val);
+
+  return db.upsert(key, val);
+}
+```
+
+```mjs displayName="storage.test.mjs"
+import assert from 'node:assert/strict';
+import { describe, it } from 'node:test';
+
+import { db } from 'db';
+
+import { save } from './storage.mjs';
+
+describe('storage', { concurrency: true }, () => {
+  it('should retrieve the requested item', async () => {
+    await db.upsert('good', 'item'); // give it something to read
+    await db.upsert('erroneous', 'item'); // give it a chance to fail
+
+    const results = await read('a', true);
+
+    assert.equal(items.length, 1); // ensure read did not retrieve erroneous item
+
+    assert.deepEqual(results[0], { key: 'good', val: 'item' });
```

Comment:
What is "items"? Did you mean:
```suggestion
    assert.equal(results.length, 1); // ensure read did not retrieve erroneous item

    assert.deepEqual(results[0], { key: 'good', val: 'item' });
```

Or even better:
```suggestion
    assert.deepEqual(results, [{ key: 'good', val: 'item' }]);
```

Reply:
I agree the second one is better for an actual implementation; I split them up here to better demonstrate what and why.